File: ai/BotSavesPrincess2.py
#!/usr/bin/python
def nextMove(n,row,column,grid):
    count = 0
    for gridRow in grid:
        princess = gridRow.find('p')

        if princess != -1:
            if count < row:
                nextMove = 'UP'
            elif count == row:
                nextMove = 'LEFT' if princess < column else 'RIGHT'
            elif count > row:
                nextMove = 'DOWN'
        
        count += 1

    return nextMove

n = int(input())
row,column = [int(i) for i in input().strip().split()]
grid = []

for i in range(0, n):
    grid.append(input())

print(nextMove(n,row,column,grid))

File: ai/BotSavesPrincess2.js
function processData(input) {
        
    var lines = input.split('\n');
    var dimension = parseInt(lines[0]);
    var mario = lines[1].split(' ');
    var marioRow = mario[0];
    var marioCol = mario[1];
    
    var nextMove = 'UNKNOWN';
    
    for(var i = 2; i <= dimension + 1; ++i){
        var gridRow = i - 2;
        var princess = lines[i].indexOf('p');
        
        if(princess !== -1){
            if(gridRow < marioRow){
                nextMove = 'UP';
            }else if(gridRow == marioRow){
                nextMove = princess < marioCol ? 'LEFT' : 'RIGHT';
            }else if(gridRow > marioRow){
                nextMove = 'DOWN';
            }
        }
    }

    process.stdout.write(nextMove);
}

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
   processData(_input);
});
